src/spreadinterp.cpp

```diff
@@ -2047,254 +1610,658 @@ void get_subgrid(BIGINT &offset1, BIGINT &offset2, BIGINT &offset3, BIGINT &padd
+           (long long)M, (long long)N1, (long long)N2, (long long)N3, nthr);
+  timer.start();
+  std::fill(data_uniform, data_uniform + 2 * N, 0.0); // zero the output array
+  if (opts.debug) printf("\tzero output array\t%.3g s\n", timer.elapsedsec());
+  if (M == 0)                                         // no NU pts, we're done
+    return 0;
 
-template<uint8_t ns, uint8_t kerevalmeth, class T, class simd_type, typename... V>
-auto ker_eval(FLT *FINUFFT_RESTRICT ker, const finufft_spread_opts &opts,
-              const V... elems) noexcept {
-  /* Utility function that allows to move the kernel evaluation outside the spreader for
-     clarity
-     Inputs are:
-     ns = kernel width
-     kerevalmeth = kernel evaluation method
-     T = (single or double precision) type of the kernel
-     simd_type = xsimd::batch for Horner
-     vectorization (default is the optimal simd size)
-     finufft_spread_opts as Horner needs
-     the oversampling factor
-     elems = kernel arguments
-     Examples usage is
-     ker_eval<ns,kerevalmeth>(opts, x, y, z) // for 3D or
-     ker_eval<ns, kerevalmeth>(opts, x, y) // for 2D or
-     ker_eval<ns, kerevalmeth>(opts, x) // for 1D
-   */
-  const std::array inputs{elems...};
-  // compile time loop, no performance overhead
-  for (auto i = 0; i < sizeof...(elems); ++i) {
-    // compile time branch no performance overhead
-    if constexpr (kerevalmeth == 1) {
-      if (opts.upsampfac == 2.0) {
-        eval_kernel_vec_Horner<ns, 200, simd_type>(ker + (i * MAX_NSPREAD), inputs[i],
-                                                   opts);
-      }
-      if (opts.upsampfac == 1.25) {
-        eval_kernel_vec_Horner<ns, 125, simd_type>(ker + (i * MAX_NSPREAD), inputs[i],
-                                                   opts);
-      }
+  auto spread_single = (nthr == 1) || (M * 100 < N); // low-density heuristic?
+  spread_single      = false;                        // for now
+  timer.start();
+  if (spread_single) { // ------- Basic single-core t1 spreading ------
+    for (UBIGINT j = 0; j < M; j++) {
+      // *** todo, not urgent
+      // ... (question is: will the index wrapping per NU pt slow it down?)
     }
-    if constexpr (kerevalmeth == 0) {
-      alignas(simd_type::arch_type::alignment()) std::array<T, MAX_NSPREAD> kernel_args{};
-      set_kernel_args<ns>(kernel_args.data(), inputs[i]);
-      evaluate_kernel_vector<ns>(ker + (i * MAX_NSPREAD), kernel_args.data(), opts);
+    if (opts.debug) printf("\tt1 simple spreading:\t%.3g s\n", timer.elapsedsec());
+  } else { // ------- Fancy multi-core blocked t1 spreading ----
+           // Splits sorted inds (jfm's advanced2), could double RAM.
+    // choose nb (# subprobs) via used nthreads:
+    auto nb = std::min((UBIGINT)nthr, M); // simply split one subprob per thr...
+    if (nb * (BIGINT)opts.max_subproblem_size < M) { // ...or more subprobs to cap size
+      nb = 1 + (M - 1) / opts.max_subproblem_size;   // int div does
```

Comment:
(M + opts.max_subproblem_size -1) /opts.max_subproblem_size is the standard way.

Reply:
This makes no difference. I coded it the way it is..